fix: module imports should not use type index as their function identifier

ColinEberhardt opened the pull request:
fixes #406 

@xtuc there is a wasm-edit test failing as a result of this change, wondering if you have any pointers?

ColinEberhardt commented on packages/wasm-parser/test/fixtures/import/complex/actual.wat:
```diff
@@ -0,0 +1,5 @@
+(module
```
this simple test, which adds a couple of type definitions, was enough to repro the issues from #406

ColinEberhardt commented on packages/wast-parser/src/grammar.js:
```diff
@@ -419,6 +417,8 @@ export function parse(tokensList: Array<Object>, source: string): Program {
         const fnParams = [];
         const fnResult = [];
 
+        let fnName = t.identifier(getUniqueName("func"));
```
this change was not necessarily needed to make tests pass, but this does mean we have the same generated names when parsing wat

xtuc replied:
Following https://github.com/xtuc/webassemblyjs/pull/411/files#r202406723 we could add our own func increment, or maybe we already have this information?

xtuc commented:
The "AST synchronization" test suite in `wasm-edit` ensure that we mutate the AST correctly by comparing it to a freshly decoded AST.

xtuc commented on packages/wasm-parser/src/decoder.js:
```diff
@@ -432,7 +432,7 @@ export function decode(ab: ArrayBuffer, opts: DecoderOpts): Program {
         }
 
-        const id = t.numberLiteralFromRaw(typeindex);
+        const id = getUniqueName("func");
```
Maybe we should try to eliminate the unique name generator (because statefull), and rather use the `typeindex` here.

ColinEberhardt commented:
> The "AST synchronization" test suite in wasm-edit ensure that we mutate the AST correctly by comparing it to a freshly decoded AST.

@xtuc ok, I've had a look at the test and understand what it is doing now. I've fixed the test by making the following changes:

- the `makeFuncNodes` function was adding the AST for the function type as well as the function itself. I've moved the construction of function types into a separate step as the module imports use this too
- I've changed the order, to add the module import first
- following this, the test adds a function, which now has an index of 1, and an identifier of `func_1`. As discussed in https://github.com/xtuc/webassemblyjs/pull/409 - this is something that we'll probably remove in future, but for now, this fixes the original issue where type indices were being used as function indices.

ColinEberhardt commented:
The tests are currently failing, but looks like a CI issue rather than a unit test failure!

ColinEberhardt commented:
kicked Travis and the build works now :-)
